File: archivelib-sys2/c-lib/src/r_compress/fn211.cpp

#include "r_compress.hpp"

int RCompress::fn211(int _212, uint16_t *_213, uint8_t *_214, uint16_t *_215) {
  int _226, _276, _289, _292;
  int16_t _227;
  data->dat174 = (int16_t)_212;
  data->dat_arr_cursor187 = _213;
  data->dat_arr_cursor178 = _214;
  _292 = data->dat174;
  _227 = 0;
  data->dat_arr177[1] = 0;
  for (_226 = 0; _226 < data->dat174; _226++) {
    data->dat_arr_cursor178[_226] = 0;
    if (data->dat_arr_cursor187[_226])
      data->dat_arr177[++_227] = (int16_t)_226;
  }
  if (_227 < 2) {
    _215[data->dat_arr177[1]] = 0;
    return data->dat_arr177[1];
  }
  for (_226 = _227 / 2; _226 >= 1; _226--)
    fn225(_226, data->dat_arr_cursor187, data->dat_arr177, _227);
  data->dat_arr_cursor188 = _215;
  do {
    _226 = data->dat_arr177[1];
    if (_226 < data->dat174)
      *data->dat_arr_cursor188++ = (uint16_t)_226;
    data->dat_arr177[1] = data->dat_arr177[_227--];
    fn225(1, data->dat_arr_cursor187, data->dat_arr177, _227);
    _276 = data->dat_arr177[1];
    if (_276 < data->dat174)
      *data->dat_arr_cursor188++ = (uint16_t)_276;
    _289 = _292++;
    data->dat_arr_cursor187[_289] = (uint16_t)(data->dat_arr_cursor187[_226] +
                                               data->dat_arr_cursor187[_276]);
    data->dat_arr177[1] = (int16_t)_289;
    fn225(1, data->dat_arr_cursor187, data->dat_arr177, _227);
    data->dat_arr189[_289] = (uint16_t)_226;
    data->dat_arr190[_289] = (uint16_t)_276;
  } while (_227 > 1);
  data->dat_arr_cursor188 = _215;
  fn228(_289);
  fn230(_212, _214, _215);
  return _289;
}

File: archivelib-rs-tdd/src/compress/fn211.rs
use crate::compress::{CompressU16ArrayAlias, CompressU8ArrayAlias, RCompressData, Result};

use crate::support::ArrayAlias;
use std::io::{Read, Write};

impl<R: Read, W: Write> RCompressData<R, W> {
  pub fn fn211(
    &mut self,
    var212: i32,
    dat_arr_cursor187: &mut CompressU16ArrayAlias<'_>,
    dat_arr_cursor178: &mut CompressU8ArrayAlias<'_>,
    dat_arr_cursor188: &mut CompressU16ArrayAlias<'_>,
  ) -> Result<i32> {
    let dat_arr_cursor178_offset = dat_arr_cursor178.offset(self);
    let dat_arr_cursor188_offset = dat_arr_cursor188.offset(self);

    self.dat174 = var212 as i16;
    let mut var292 = var212;
    let mut var227 = 0;
    self.dat_arr177[1] = 0;
    for i in 0..(self.dat174 as usize) {
      dat_arr_cursor178.set(self, i, 0);
      if 0 != dat_arr_cursor187.get(self, i) {
        var227 += 1;
        self.dat_arr177[var227] = i as i16
      }
    }
    if var227 < 2 {
      dat_arr_cursor188.set(self, self.dat_arr177[1] as usize, 0);
      return Ok(self.dat_arr177[1] as i32);
    } else {
      let mut run_start226 = (var227 / 2) as i16;
      while run_start226 >= 1 {
        self.fn225(run_start226 as i32, dat_arr_cursor187, var227 as i16);
        run_start226 -= 1
      }
      let mut var289;
      loop {
        run_start226 = self.dat_arr177[1];
        if run_start226 < self.dat174 {
          dat_arr_cursor188.set(self, 0, run_start226 as u16);
          dat_arr_cursor188.shift(self, 1);
        }
        self.dat_arr177[1] = self.dat_arr177[var227];
        var227 -= 1;
        self.fn225(1, dat_arr_cursor187, var227 as i16);
        let run_length276 = self.dat_arr177[1];
        if run_length276 < self.dat174 {
          dat_arr_cursor188.set(self, 0, run_length276 as u16);
          dat_arr_cursor188.shift(self, 1);
        }
        var289 = var292;
        var292 = var292 + 1;
        dat_arr_cursor187.set(
          self,
          var289 as usize,
          dat_arr_cursor187.get(self, run_start226 as usize)
            + dat_arr_cursor187.get(self, run_length276 as usize),
        );
        self.dat_arr177[1] = var289 as i16;
        self.fn225(1, dat_arr_cursor187, var227 as i16);
        self.dat_arr189[var289 as usize] = run_start226 as u16;
        self.dat_arr190[var289 as usize] = run_length276 as u16;
        if !(var227 > 1) {
          break;
        }
      }
      dat_arr_cursor188.set_offset(self, dat_arr_cursor188_offset);
      self.fn228(var289, dat_arr_cursor178, dat_arr_cursor188);
      dat_arr_cursor188.set_offset(self, dat_arr_cursor188_offset);
      dat_arr_cursor178.set_offset(self, dat_arr_cursor178_offset);
      self.fn230(var212, dat_arr_cursor178, dat_arr_cursor188);
      return Ok(var289);
    };
  }
}
